### PR TITLE
fix(versioner): fix perms of created folders (fixes #9626)

### LINKED ISSUE #9626: Files “created” by Syncthing ignore UMask (.stfolder and .stversions)
### What happened?

I am using Syncthing as a `systemd` service on a Raspberry Pi with Raspbian. I have set it up so it’s run by a `syncthing` user, but to keep access to all users in the group I’ve set the `UMask` parameter to `0002` in the `syncthing@syncthing.service` and for good measure in the `.bash_rc`, `.bash_pofile` and `.profile` files for that user.

Setting `umask` in the service file works correctly for synchronization, but Syncthing seems to ignore it when the files are “created” by Syncthing itself and not synchronized. I’ve noticed this happening both in `.stfolder` and subfiles (when a new synchronization folder is created) and the `.stversions` and subfolders (only directories, not files).

E.g.: Creating a new, still unshared folder through the web GUI:

```
pi@raspberrypi:/media/data/Sync $ ls TEST/ -ld
drwxrwsr-x 3 syncthing drive-users 4096 Aug  3 00:10 TEST/
pi@raspberrypi:/media/data/Sync $ ls TEST/ -lAR
TEST/:
total 4
drwxr-sr-x 2 syncthing drive-users 4096 Aug  3 00:10 .stfolder

TEST/.stfolder:
total 4
-rw-r--r-- 1 syncthing drive-users 117 Aug  3 00:10 syncthing-folder-ccaee1.txt
```

As for `.stversions`, I use “Simple” file versioning, and when I remove files the `.stversions` folder and any subfolders are created without group writing permissions, seemingly ignoring the `umask`. The files keep the writing permission, but without the folder writing permission I can’t delete them.

E.g.: Note that in this folder I have turned on “Ignore Permissions”. Also, in this case `.stfolder` has group write permission set manually.

```
pi@raspberrypi:/media/data/Sync/Linux-Documents $ ls -lAR
.:
total 8
drwxrwsr-x 2 syncthing drive-users 4096 Aug  2 23:54 'New Folder'
drwxrwsr-x 2 syncthing drive-users 4096 May  7 22:56  .stfolder

'./New Folder':
total 4
-rw-rw-r-- 1 syncthing drive-users 2 Aug  2 23:54 'Text File.txt'

./.stfolder:
total 0
pi@raspberrypi:/media/data/Sync/Linux-Documents $ ls -lAR
.:
total 8
drwxrwsr-x 2 syncthing drive-users 4096 May  7 22:56 .stfolder
drwxr-sr-x 3 syncthing drive-users 4096 Aug  2 23:55 .stversions

./.stfolder:
total 0

./.stversions:
total 4
drwxr-sr-x 2 syncthing drive-users 4096 Aug  2 23:55 'New Folder'

'./.stversions/New Folder':
total 4
-rw-rw-r-- 1 syncthing drive-users 2 Aug  2 23:54 'Text File~20240802-235507.txt'
pi@raspberrypi:/media/data/Sync/Linux-Documents $ rm .stversions/New\ Folder/Text\ File~20240802-235507.txt
rm: cannot remove '.stversions/New Folder/Text File~20240802-235507.txt': Permission denied
```

For more information, here is my service file:

```
$ cat /lib/systemd/system/syncthing@syncthing.service 
[Unit]
Description=Syncthing - Open Source Continuous File Synchronization for %I
Documentation=man:syncthing(1)
After=network.target

[Service]
User=%i
ExecStart=/usr/bin/syncthing -no-browser -no-restart -logflags=0
Restart=on-failure
SuccessExitStatus=3 4
RestartForceExitStatus=3 4
UMask=0002

# Hardening
ProtectSystem=full
PrivateTmp=true
SystemCallArchitectures=native
MemoryDenyWriteExecute=true
NoNewPrivileges=true

[Install]
WantedBy=multi-user.target
```

I have brought up this problem on the forum as well, where @calmh [identified it as a bug](https://forum.syncthing.net/t/files-created-by-syncthing-ignore-umask-stfolder-and-stversions/22530/2?u=xitoshi):

>You’re right, mkdirs in the versioner code use 0755 as the base permissions, so umask won’t help there (to get group write). Please feel free to file an issue on this, and perhaps a fix :slight_smile:

>Probably the real fix would be to grab the permissions from the corresponding directory in the synced folder.

### Syncthing version

syncthing v1.27.9 "Gold Grasshopper" (go1.22.4 linux-arm) debian@github.syncthing.net 2024-07-01 03:45:16 UTC [noupgrade]

### Platform & operating system

Raspbian GNU/Linux 11 (bullseye)

### Browser version

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by ashishbhate**
### Purpose
Fixes #9626. 
As suggested in the linked issue, I've updated the versioner code to use the permissions of the corresponding directory in the synced folder, when creating the folder in the versions directory

### Testing
- Some tests are included with the PR. Happy to add more if you think there are some edge-cases that we're missing.
- I've tested manually on linux to confirm the permissions of the created directories.
- I haven't tested on Windows or OSX (I don't have access to these OS)

### CODE REVIEW

File: lib/versioner/util.go [hunk 1]
```diff
@@ -190,6 +190,53 @@ func archiveFile(method fs.CopyRangeMethod, srcFs, dstFs fs.Filesystem, filePath
...
 
+func dupDirTree(srcFs, dstFs fs.Filesystem, folderPath string) error {
+	// Return early if the folder already exists.
+	_, err := dstFs.Stat(folderPath)
+	if err == nil || !fs.IsNotExist(err) {
+		return err
+	}
+	i := 0
+	hadParent := true
+	for i < len(folderPath) {
```
Comment (calmh) [inline lib/versioner/util.go:201]: `for i := range folderPath` should be simpler here I think, avoids `i++` in the loop and avoids looking at partial code points.

(I had to think through the subtleties of string indexing and bytes vs code points for the things that happen here, but I think it's fine in both you version and in the above change.)

**Review by calmh — COMMENTED** [review]
Looks in principle reasonable

File: lib/versioner/util.go (hunk 1 above)
Comment (bt90) [reply lib/versioner/util.go:201]: Does go do some kind of constant folding or is `len(folderPath)` evaluated on each loop iteration?

Comment (calmh) [reply lib/versioner/util.go:201]: Having len(foo) as a loop condition is fine, you'd have to check the assembly output for the details but both strings and slices have the len stored as an integer in the object so even worst case it's one integer load.

(but the range construction is better anyway)

Comment (ashishbhate) [reply lib/versioner/util.go]: You're right, using a range over the string is better. It's clearer code and saves us from unnecessarily looking at partial code points. It also alleviates bt90's concerns about constant folding.

Yep, I too believe this code is safe based on how we're indexing and slicing.
There are only two path separators defined in the Go standard library `\` (windows [1])  and `/` (unix [2] and plan9 [3]). Both of these have single-byte utf8 representations.
We will only run into issues if the first byte of a multi-byte utf8 rune is the same as one of the single-byte path separators. Based on how utf8 encoding works, that is not possible [4]


[1] https://go.googlesource.com/go/+/refs/tags/go1.24.3/src/os/path_windows.go#14
[2] https://go.googlesource.com/go/+/refs/tags/go1.24.3/src/os/path_unix.go#10
[3] https://go.googlesource.com/go/+/refs/tags/go1.24.3/src/os/path_plan9.go#8
[4] https://en.wikipedia.org/wiki/UTF-8#Description

File: lib/versioner/util.go [hunk 2]
```diff
@@ -190,6 +190,50 @@ func archiveFile(method fs.CopyRangeMethod, srcFs, dstFs fs.Filesystem, filePath
...
+}
+
+func dupDirWithPerms(srcFs, dstFs fs.Filesystem, folderPath string) error {
+	srcStat, err := srcFs.Stat(folderPath)
+	if err != nil {
+		return err
+	}
+	// If we call Mkdir with srcStat.Mode(), we won't get the expected perms because of umask
+	// So, we create the folder with 0755, and then change the perms to the srcStat.Mode()
+	err = dstFs.Mkdir(folderPath, 0o755)
```
Comment (calmh) [inline lib/versioner/util.go:230]:
```suggestion
	// So, we create the folder with 0700, and then change the perms to the srcStat.Mode()
	err = dstFs.Mkdir(folderPath, 0o700)
```

avoids having (possibly) overly permissive bits for a while

**Review by calmh — APPROVED** [review]

**Comment by ashishbhate** [comment]
>Needs test fix for Windows where permissions don't work the same way. Possibly just skip it entirely there.

I don't have access to a Windows machine to test this out. I'll try setting up a VM over the weekend. If I can't figure out how to make it work, I'll skip the test.

The Ubuntu CI is failing on `TestCompletionEmptyGlobal`. I'm not sure its related, but I'll take a look at it this weekend as well.

**Comment by calmh** [comment]
I think t.Skip() is the correct choice here. The only permissions we "see" on Windows is the read-only bit, which is annoying to replicate.

The other failure is flakiness, you can ignore it.

**Comment by ashishbhate** [comment]
>I think t.Skip() is the correct choice here. The only permission we "see" on Windows is the read-only bit, which is annoying to replicate.

Done
